### PR TITLE
Coq: adapt for upcoming 8.14

### PR DESCRIPTION
**Pull request by Zimmi48**
###### Motivation for this change

This restores the ability to use the override mechanism to build the development version of Coq and should make it easier to package the upcoming Coq 8.14 version in nixpkgs.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @CohenCyril @vbgl

### CODE REVIEW

File: pkgs/applications/science/logic/coq/default.nix [hunk 1]
```diff
@@ -129,14 +129,14 @@ self = stdenv.mkDerivation {
...
+       else [ ocamlPackages.lablgtk ])
+    ++ optional (versionAtLeast "8.14") [ ocamlPackages.dune_2 ];
 
   postPatch = ''
     UNAME=$(type -tp uname)
     RM=$(type -tp rm)
-    substituteInPlace configure --replace "/bin/uname" "$UNAME"
     substituteInPlace tools/beautify-archive --replace "/bin/rm" "$RM"
-    substituteInPlace configure.ml --replace '"md5 -q"' '"md5sum"'
+    substituteInPlace ${if versionAtLeast "8.14" then "tools/configure/" else ""}configure.ml --replace '"md5 -q"' '"md5sum"'
```
Comment (vbgl) [inline pkgs/applications/science/logic/coq/default.nix:139]: Are there still references to `md5` in the `configure.ml` of 8.14?

File: pkgs/applications/science/logic/coq/default.nix [hunk 2]
```diff
@@ -129,14 +129,14 @@ self = stdenv.mkDerivation {
     ++ optionals buildIde
       (if versionAtLeast "8.10"
        then [ ocamlPackages.lablgtk3-sourceview3 glib gnome.adwaita-icon-theme wrapGAppsHook ]
-       else [ ocamlPackages.lablgtk ]);
+       else [ ocamlPackages.lablgtk ])
+    ++ optional (versionAtLeast "8.14") [ ocamlPackages.dune_2 ];
```
Comment (vbgl) [inline pkgs/applications/science/logic/coq/default.nix:133]: Please move the semicolon to the next line.

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 128603` at 53c902b1 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>40 packages skipped due to time constraints:</summary>
  <ul>
    <li>compcert (coqPackages.compcert)</li>
    <li>coqPackages.CoLoR</li>
    <li>coqPackages.ITree</li>
    <li>coqPackages.QuickChick</li>
    <li>coqPackages.VST</li>
    <li>coqPackages.Verdi</li>
    <li>coqPackages.coq-bits</li>
    <li>coqPackages.coq-elpi</li>
    <li>coqPackages.coqeal</li>
    <li>coqPackages.coqhammer</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>29 packages built successfully:</summary>
  <ul>
    <li>coq (coqPackages.coq ,coq_8_13)</li>
    <li>coqPackages.Cheerios</li>
    <li>coqPackages.InfSeqExt</li>
    <li>coqPackages.StructTact</li>
    <li>coqPackages.aac-tactics</li>
    <li>coqPackages.autosubst</li>
    <li>coqPackages.bignums</li>
    <li>coqPackages.coq-ext-lib</li>
    <li>coqPackages.dpdgraph</li>
    <li>coqPackages.equations</li>
    <li>coqPackages.mathcomp-algebra</li>
    <li>coqPackages.mathcomp-bigenough</li>
    <li>coqPackages.mathcomp-fingroup</li>
    <li>coqPackages.mathcomp-finmap</li>
    <li>coqPackages.ssreflect (coqPackages.mathcomp-ssreflect)</li>
    <li>coqPackages.mathcomp-zify</li>
    <li>coqPackages.multinomials</li>
    <li>coqPackages.paco</li>
    <li>coqPackages.paramcoq</li>
    <li>coqPackages.pocklington</li>
    <li>coqPackages.reglang</li>
    <li>coqPackages.relation-algebra</li>
    <li>coqPackages.semantics</li>
    <li>coqPackages.simple-io</li>
    <li>coqPackages.topology</li>
    <li>coqPackages.zorns-lemma</li>
    <li>coq_8_5</li>
    <li>coq_8_7</li>
    <li>coq_8_8</li>
  </ul>
</details>

File: pkgs/applications/science/logic/coq/default.nix (hunk 1 above)
Comment (Zimmi48) [reply pkgs/applications/science/logic/coq/default.nix:139]: Good point, this was removed in 8.7.

**Review by vbgl — APPROVED** [review]
May I squash all these changes into one commit or should the intermediate steps be preserved?

**Comment by Zimmi48** [comment]
Indeed, please squash :)

File: pkgs/applications/science/logic/coq/default.nix [hunk 3]
```diff
@@ -129,14 +129,15 @@ self = stdenv.mkDerivation {
     ++ optionals buildIde
       (if versionAtLeast "8.10"
        then [ ocamlPackages.lablgtk3-sourceview3 glib gnome.adwaita-icon-theme wrapGAppsHook ]
-       else [ ocamlPackages.lablgtk ]);
+       else [ ocamlPackages.lablgtk ])
+    ++ optional (versionAtLeast "8.14") [ ocamlPackages.dune_2 ]
```
Comment (CohenCyril) [inline pkgs/applications/science/logic/coq/default.nix:133]: Should have been `optional (versionAtLeast "8.14") ocamlPackages.dune_2`

Comment (CohenCyril) [reply pkgs/applications/science/logic/coq/default.nix:133]: @Zimmi48

Comment (Zimmi48) [reply pkgs/applications/science/logic/coq/default.nix:133]: I've opened #128747.

**Pull request merged**